### PR TITLE
Remove auto increment from jobs table

### PR DESCRIPTION
**Pull request by iwiznia**
@tylerkaraszewski, @mea36 please review.

Part of https://github.com/Expensify/Expensify/issues/75339

Tests:
- Start bedrock with this changes.
- Run [this migration](https://github.com/Expensify/Expensify/issues/75339#issuecomment-374274942).
- Start bedrock.
- Create some jobs, check the IDs returned (on new DB or a one with an empty jobs table it starts at 1).
- Restart bedrock.
- Create another job, check it uses the correct ID

### CODE REVIEW

File: plugins/Jobs.cpp [hunk 1]
```diff
@@ -345,6 +358,14 @@ bool BedrockPlugin_Jobs::processCommand(SQLite& db, BedrockCommand& command) {
     STable& content = command.jsonContent;
     const string& requestVerb = request.getVerb();
 
+    if (!lastJobID) {
```
Comment (iwiznia) [inline plugins/Jobs.cpp]: I don't think this is the best place for this. Initially I wanted to add it to the `initialize` method, but I did not have access to the db object there... let me know if there's a better place to put this.

Comment (tylerkaraszewski) [reply plugins/Jobs.cpp]: What about doing it in `upgradeDatabase`?

Comment (iwiznia) [reply plugins/Jobs.cpp]: I thought about it, seems just as misplaced as here, but I can move it there if you think is better.

File: plugins/Jobs.h [hunk 2]
```diff
@@ -11,6 +11,8 @@ class BedrockPlugin_Jobs : public BedrockPlugin {
     virtual bool processCommand(SQLite& db, BedrockCommand& command);
 
   private:
+    int64_t lastJobID;
```
Comment (iwiznia) [inline plugins/Jobs.h]: Heh, reviewing my own code... I thought that this is actually wrong as it's not thread safe, right? Should I use a `std::atomic_int`?

File: plugins/Jobs.cpp (hunk 1 above)
Comment (tylerkaraszewski) [reply plugins/Jobs.cpp]: At least there it only gets run once at startup instead of on every request.

File: plugins/Jobs.h (hunk 2 above)
Comment (tylerkaraszewski) [reply plugins/Jobs.h]: Depends on how many threads can change it.

**Comment by cead22** [comment]
Talking about this with Ionatan I don't think we should pass the current jobID as a cli param when running webrock because the problem this is trying to solve is extremely unlikely and thus we shouldn't spend time or code fixing a problem that we won't experience.

My understanding is that in this following scenarios we could wrongfully reuse a jobID:

 1. A job is created with retryAfter
 2. The job is pulled and marked as RUNNING
 3. Enough time passes that we'll retry the job per `retryAfter`
 4. Someone else grabs the job and finishes it at which point we delete the job
 5. The original client that pulled the job finishes it, and because we reused the jobID they finish the wrong job

For that to happen it means that we didn't create any new jobs between 2. and 5. which I don't see how it could realistically happen.

**Comment by tylerkaraszewski** [comment]
Let's wait on this until after this PR has been deployed: https://github.com/Expensify/Bedrock/pull/402

Then we can see if the change to SQLite makes this change unnecessary.

**Comment by iwiznia** [comment]
@tylerkaraszewski ready to review/merge. I disagree with waiting on #402. That is an optimization to the auto increment when you provide the ID of an auto increment table. That change will have no effect on the jobs plugins as it is, as it IS relying on auto increment.

> Talking about this with Ionatan I don't think we should pass the current jobID as a cli param when running webrock

@cead22, this is not what we are doing, you misunderstood the change. What we are doing is NOT using the auto increment from sqlite and instead generating our own jobID internally in the jobs plugin (so there's absolutely no change for the code using the jobs plugin).

**Comment by tylerkaraszewski** [comment]
I'm ok with this change now. Does anyone object to me merging it?

**Comment by mea36** [comment]
I'd prefer to merge it after the bedrock deploy so we have a week of testing this locally

**Comment by iwiznia** [comment]
I disagree, can we get it out today that's the hight risk deploy? I can do any test you ask for Mel to quail your concerns.

**Comment by tylerkaraszewski** [comment]
@mea36 brought up a good issue here with the max IDs:

1. Assume a node is started up as a slave.
2. It reads the `maxJobID` as `100`.
3. It then receives new and deleted jobs from master via synchronization. None of these get reflected in `maxJobID`.
4. Then master goes down and this node gets promoted to master. 
5. This node starts creating new jobs with IDs starting at 100.

I think all of the issues we know of so far can be solved using GUIDs in place in incrementing IDs.

**Comment by cead22** [comment]
> I'm ok with this change now. Does anyone object to me merging it?

Me. Can somebody please humor me and explain the scenario that we're trying to fix/avoid with the in-memory lastJobID?

**Comment by tylerkaraszewski** [comment]
> Can somebody please humor me and explain the scenario that we're trying to fix/avoid with the in-memory lastJobID?

If every insert requires looking at the last row in the existing table, then we have the same problem with inserts all accessing the same part of the DB that causes them to conflict that we have now.

**Comment by cead22** [comment]
👍 Thanks, I just looked at the code after seeing Ioni's note that I misunderstood the change and I can see this isn't taking the maxJobID from a cli param.

Just to clarify, we're now just querying the max job ID once and then doing lastJobID++ to generate the new ID instead of always querying the max job ID, right?

As for that scenario from Mel/Tyler, it seems that we should always get maxJobID from master right?

**Comment by tylerkaraszewski** [comment]
`maxJobID` only really matters on master. I actually think I was being overly cautious with my previous worry, but @iwiznia is going to build a test for this to make sure it does the right thing.

**Comment by iwiznia** [comment]
Yep, this _should_ work, because we run upgradeDB when switching to MASTERING [here](https://github.com/Expensify/Bedrock/blob/master/BedrockServer.cpp#L257-L258)... but I need to test this properly first.

**Comment by iwiznia** [comment]
What?! Why the build is failing if no tests failed?!?!

**Comment by iwiznia** [comment]
ok, tests pass now 😄

**Comment by iwiznia** [comment]
Bump?

**Comment by iwiznia** [comment]
Double bump!

**Pull request merged**